### PR TITLE
Remove some common classes, spread helper functions to where they are used

### PR DESCRIPTION
**Pull request by timtay-microsoft**
Several classes in the common folder were only used by one library, so those classes have been moved to that library.

Some classes in the common folder had functions used only by one library and other functions used by a different library. Those functions have been split from the common class and moved to the library where they are used.

This is just a first step to removing the common folder, so we'll have more PRs like this in the future

### CODE REVIEW

File: iothub/device/src/Common/Amqp/ClientWebSocketTransport.cs [hunk 1]
```diff
@@ -341,5 +341,24 @@ private void ThrowIfNotOpen()
 
             throw new AmqpException(AmqpErrorCode.IllegalState, null);
         }
+
+        private static void ValidateBufferBounds(byte[] buffer, int offset, int size)
+        {
+            if (buffer == null)
+            {
+                throw new ArgumentNullException(nameof(buffer));
+            }
```
Comment (drwill-ms) [inline iothub/device/src/Common/Amqp/ClientWebSocketTransport.cs:350]: `Argument.AssertNotNull()`?

File: iothub/device/src/DirectMethod/MethodResponse.cs [hunk 2]
```diff
@@ -44,12 +44,7 @@ public MethodResponse(int status)
             Justification = "Cannot change property types on public classes.")]
         public byte[] Result
         {
-            private set
-            {
-                Utils.ValidateDataIsEmptyOrJson(value);
-
-                _result = value;
-            }
+            private set => _result = value;
```
Comment (drwill-ms) [inline iothub/device/src/DirectMethod/MethodResponse.cs:47]: Odd that set is above get. Mind swapping them?

File: iothub/device/src/Transport/Mqtt/MqttIotHubAdapter.cs [hunk 3]
```diff
@@ -1395,6 +1397,33 @@ internal static string PopulateMessagePropertiesFromMessage(string topicName, Me
             return msg;
         }
 
+        private static IReadOnlyDictionary<TKey, TValue> MergeDictionaries<TKey, TValue>(IDictionary<TKey, TValue>[] dictionaries)
+        {
+            // No item in the array should be null.
+            if (dictionaries == null || dictionaries.Any(item => item == null))
+            {
+                throw new ArgumentNullException(nameof(dictionaries), "Provided dictionaries should not be null");
+            }
```
Comment (drwill-ms) [inline iothub/device/src/Transport/Mqtt/MqttIotHubAdapter.cs:1406]: I think there is an option for this in `Arguments.RequireNotNullOrEmpty<T>()` that I added to the device's Arguments class you could copy to the service one. The source for this was Azure.Core.

File: iothub/device/src/Transport/Mqtt/MqttIotHubAdapter.cs [hunk 4]
```diff
@@ -1395,6 +1397,33 @@ internal static string PopulateMessagePropertiesFromMessage(string topicName, Me
...
+
+            var result = dictionaries.SelectMany(dict => dict)
+                .ToLookup(pair => pair.Key, pair => pair.Value)
+                .ToDictionary(group => group.Key, group => group.First());
+
+            return new ReadOnlyDictionary<TKey, TValue>(result);
+        }
+
+        private static DeliveryAcknowledgement ConvertDeliveryAckTypeFromString(string value)
```
Comment (drwill-ms) [inline iothub/device/src/Transport/Mqtt/MqttIotHubAdapter.cs:1415]: We should consider making the enum deserializable so we don't have to maintain this mapping outside of the enum.

File: iothub/service/src/Http/HttpClientHelper.cs [hunk 5]
```diff
@@ -885,5 +886,13 @@ internal static HttpMessageHandler CreateDefaultHttpMessageHandler(IWebProxy web
 
             return httpMessageHandler;
         }
+
+        private static string GetClientVersion()
```
Comment (drwill-ms) [inline iothub/service/src/Http/HttpClientHelper.cs:890]: I don't see the point of decentralizing this when we now have 2 copies of the same code.

File: iothub/service/src/Http2/HttpRequestMessageFactory.cs [hunk 6]
```diff
@@ -74,5 +75,13 @@ internal HttpRequestMessage CreateRequest(HttpMethod method, Uri relativeUri, Io
 
             return message;
         }
+
+        private static string GetClientVersion()
```
Comment (drwill-ms) [inline iothub/service/src/Http2/HttpRequestMessageFactory.cs:79]: 3 copies now.

File: iothub/device/src/Transport/Mqtt/MqttIotHubAdapter.cs (hunk 4 above)
Comment (timtay-microsoft) [reply iothub/device/src/Transport/Mqtt/MqttIotHubAdapter.cs:1415]: This class will disappear once we move to MQTTNet, so I'm not going to bother here

**Review by drwill-ms — CHANGES_REQUESTED** [review]
I see value in some of these changes, but question the DRY violations caused by others.

File: iothub/device/src/Transport/Mqtt/MqttIotHubAdapter.cs (hunk 3 above)
Comment (timtay-microsoft) [reply iothub/device/src/Transport/Mqtt/MqttIotHubAdapter.cs:1406]: This class will disappear once we move to MQTTNet

File: iothub/service/src/Utilities/Utils.cs [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+            if (offset < 0 || offset > buffer.Length || size <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(offset));
+            }
+
+            int remainingBufferSpace = buffer.Length - offset;
+            if (size > remainingBufferSpace)
+            {
+                throw new ArgumentOutOfRangeException(nameof(size));
+            }
```
Comment (drwill-ms) [inline iothub/service/src/Utilities/Utils.cs:32]: Would this fit better in Arguments.cs?

**Review by drwill-ms — APPROVED** [review]

File: iothub/service/src/Utilities/Utils.cs (hunk 7 above)
Comment (timtay-microsoft) [reply iothub/service/src/Utilities/Utils.cs:32]: Good idea

**Pull request merged**